client/powerline.c

```diff
@@ -90,8 +91,18 @@ int main(int argc, char *argv[]) {
 	memset(&server, 0, sizeof(struct sockaddr_un));
 	server.sun_family = AF_UNIX;
 	strncpy(server.sun_path A, address, strlen(address));
+#ifdef __linux__
+	/* We use a socket in the abstract namespace on Linux, and because SUN_LEN uses strlen, it does not work for these. The struct size also fails, so compute manually. */
+	socklen = sizeof(server.sun_family) + strlen(address) A;
+#elif defined SUN_LEN
```

Comment:
> My headers (Gentoo linux) do contain SUN_LEN. But this is not going to be ever used because of #ifdef __linux__ being the first.

Correct. I haven't looked at the Gentoo headers, but if they are similar to Cent, `SUN_LEN` is broken for abstract namespace paths. I would like to keep the behavior for Linux the same, because it's currently working fine.

My `sys/un.h` header on OS X has this guard:
```c
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
/* actual length of an initialized sockaddr_un */
#define SUN_LEN(su) \
	(sizeof(*(su)) - sizeof((su)->sun_path) + strlen((su)->sun_path))
#endif	/* (!_POSIX_C_SOURCE || _DARWIN_C_SOURCE) */
```
So if we are not compiling with `_POSIX_C_SOURCE`, `SUN_LEN` should be defined.